### PR TITLE
[TEAM2-377] EIP-4339 Detect Merge With Flag

### PR DESCRIPTION
**Pull request by derHowie**
## What changed (plus any additional context for devs)
I'm throwing up a draft in preparation for the Görli merge 08/10 @ 2pm EST. I've left Ropsten in the app for now as it is a good test case for toggling between networks that have and have not merged.

cc @estebanmino

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels? (`team1/team2`, `critical path`, `release`, `dev QA`)
- [x] Did you test both iOS and Android?
- [x] If your changes are visual, did you check both the light and dark themes?
- [x] Added e2e tests? If not, please specify why
- [x] If you added new critical path files, did you update the CODEOWNERS file?
- [x] If no `dev QA` label, did you add the PR to the QA Queue?

### CODE REVIEW

File: src/handlers/web3.ts [hunk 1]
```diff
@@ -833,3 +836,25 @@ export const estimateGasLimit = async (
...
+  logger.debug('_DIFFICULTY: ', _difficulty.toString());
+  const currentHasMerged = getHasMerged();
+  logger.debug('CURRENT HAS MERGED: ', currentHasMerged);
+  if (_difficulty.toString() === '0') {
+    setHasMerged();
+  } else if (currentHasMerged) {
+    setHasNotMerged();
+  }
+  logger.debug('GET HAS MERGED FINAL: ', getHasMerged());
+};
```
Comment (derHowie) [inline src/handlers/web3.ts:860]: This works nicely when toggling between mainnet and ropsten.

There are 2 outstanding questions with regard to the flag:
1. Does it make sense to make a hook for this? So far I cannot think of a reason why the view layer will need to access this flag.
2. Should we be polling here instead?

File: src/handlers/web3.ts [hunk 2]
```diff
@@ -833,3 +836,62 @@ export const estimateGasLimit = async (
...
+  getSetMergeStorageKey(true)(network);
+};
+export const setHasNotMerged = (network: Network) => {
+  getSetMergeStorageKey(false)(network);
+};
+export const checkForTheMerge = async (
+  provider: JsonRpcProvider,
+  network: Network
+) => {
+  const block = await provider.getBlock('latest');
```
Comment (brunobar79) [inline src/handlers/web3.ts:885]: I'd only run this logic if `getHasMerged(network) !== true`

**Review by brunobar79 — APPROVED** [review]

File: src/handlers/web3.ts [hunk 3]
```diff
@@ -833,3 +836,64 @@ export const estimateGasLimit = async (
     return estimateGas(estimateGasData, provider);
   }
 };
+
+const HAS_MERGED_DEFAULTS: { [key: string]: boolean } = {
+  [Network.mainnet]: false,
+  [Network.goerli]: false,
+  [Network.ropsten]: false,
+  [Network.rinkeby]: false,
+  [Network.kovan]: false,
```
Comment (estebanmino) [inline src/handlers/web3.ts:845]: should we remove this deprecated testnets?

File: src/handlers/web3.ts [hunk 4]
```diff
@@ -833,3 +836,64 @@ export const estimateGasLimit = async (
...
+};
+export const checkForTheMerge = async (
+  provider: JsonRpcProvider,
+  network: Network
+) => {
+  const currentHasMerged = getHasMerged(network);
+  logger.debug('CURRENT HAS MERGED: ', currentHasMerged);
+  if (currentHasMerged !== true) {
+    const block = await provider.getBlock('latest');
+    logger.debug('LATEST BLOCK: ', block);
```
Comment (estebanmino) [inline src/handlers/web3.ts:889]: this one is a bit noisy when starting the app on start, did you want to see an specific param of the block?

File: src/redux/explorer.ts [hunk 5]
```diff
@@ -578,6 +582,7 @@ export const explorerInit = () => async (
   // if we're not on mainnnet
   const provider = await getProviderForNetwork(network);
   const providerUrl = provider?.connection?.url;
+  await checkForTheMerge(provider, network);
```
Comment (estebanmino) [inline src/redux/explorer.ts:585]: do we need to `await`?

**Review by estebanmino — APPROVED** [review]
amazing! working as expected while switching networks and when the merge happened there are no new request! 🐼

**Pull request merged**